### PR TITLE
add secret option to flag

### PR DESCRIPTION
**Pull request by xilabao**
To resolve the issue of security(pr #35030 ),

> @smarterclayton commented 5 days ago
> This is unfortunately not all flags that could be secrets. The best option would be to add support in spf13/pflag to tag a flag as a secret, and then use that bit to determine the list.
> 
> Also, Command() could be used in contexts that need exact parameters (for subshell execution), so we would need to add a new method or extend the signature here to allow exact flags to be retrieved.

we could add a secret option to the flags.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35408)

### CODE REVIEW

**Comment by smarterclayton** [comment]
@kubernetes/kubectl if someone can take a pass

**Comment by fabianofranz** [comment]
@xilabao did you open a PR to https://github.com/spf13/pflag proposing the change?

**Comment by xilabao** [comment]
@fabianofranz  I didn't open a PR to https://github.com/spf13/pflag, should I do that frist?

**Comment by fabianofranz** [comment]
@xilabao we update our dependencies from time to time, so in case your changes are not in the main dependency repo, it would get lost in the next update.  ;)

**Comment by xilabao** [comment]
@smarterclayton @fabianofranz spf13/pflag don't want to add secret option, Could we add secret label to the end of description?

**Comment by smarterclayton** [comment]
@eparis since he thought there was a way to do this without changes to pflag.

The "right" structure for this is to associate flags with secrecy, and have a method that can correlate the secrecy to generate a command line (as noted by eric) for use.  This isn't really specific to kubectl since some of our other commands may wish to log the flags they are started with (same problem). I'd prefer the code not be deeply tied to anything in Kube.

**Comment by eparis** [comment]
@smarterclayton I left 2 paths forward.

1. Add a new function to cobra, something like `GetCmdLine` which did its best to recreate the command line based on cobra + flags + remaining args. While we couldn't maintain the order of flags, it would be possible. If we did that I agreed that something (likely cobra) could add "secret hiding".
2. The other option is to continue to 'reparse' the os.Args in kubernetes. If we are going to use os.Args in kubernetes then cobra/pflag don't need to change. kubernetes needs to track secret flags and when kubernetes uses os.Args it needs to do that substitution. (tracking of 'secret flags' can already be accomplished via annotations)

Number 2 is much smaller and easier. Number 1 is generic and useful for all users of cobra. I think either is reasonable. Of course I'd rather see the first, but I can't expect everyone to solve everyone else's problems when they only want to scratch their own itch.

I'd be glad to talk about how I think a good general solution might work to anyone interested!

**Comment by eparis** [comment]
I'm making a comment just so the bot thinks things are happening.

File: pkg/client/unversioned/clientcmd/overrides.go [hunk 1]
```diff
@@ -133,39 +147,39 @@ func RecommendedConfigOverrideFlags(prefix string) ConfigOverrideFlags {
...
 	return ClusterOverrideFlags{
-		APIServer:             FlagInfo{prefix + FlagAPIServer, "", "", "The address and port of the Kubernetes API server"},
-		APIVersion:            FlagInfo{prefix + FlagAPIVersion, "", "", "DEPRECATED: The API version to use when talking to the server"},
-		CertificateAuthority:  FlagInfo{prefix + FlagCAFile, "", "", "Path to a cert. file for the certificate authority"},
-		InsecureSkipTLSVerify: FlagInfo{prefix + FlagInsecure, "", "false", "If true, the server's certificate will not be checked for validity. This will make your HTTPS connections insecure"},
+		APIServer:             FlagInfo{false, prefix + FlagAPIServer, "", "", "The address and port of the Kubernetes API server"},
+		APIVersion:            FlagInfo{false, prefix + FlagAPIVersion, "", "", "DEPRECATED: The API version to use when talking to the server"},
+		CertificateAuthority:  FlagInfo{false, prefix + FlagCAFile, "", "", "Path to a cert. file for the certificate authority"},
+		InsecureSkipTLSVerify: FlagInfo{false, prefix + FlagInsecure, "", "false", "If true, the server's certificate will not be checked for validity. This will make your HTTPS connections insecure"},
```
Comment (lavalamp) [inline pkg/client/unversioned/clientcmd/overrides.go]: Adding a struct field like this really uglifies this code. Is it actually worth it?

**Comment by lavalamp** [comment]
looking at the source PR, I don't see how this could possibly work. We have no way of knowing which ENV vars or flags passed to the container are sensitive?

**Comment by xilabao** [comment]
>We have no way of knowing which ENV vars or flags passed to the container are sensitive?

pflag doesn't support it. We may add a mark on the Description. @lavalamp PTAL

**Comment by eparis** [comment]
I thought at this point the idea was to add an annotation on the flag, not to the description. Then use that annotation to hide the value when recreating the command line...

**Comment by xilabao** [comment]
Ah... I misunderstood. @eparis @lavalamp PTAL

**Comment by eparis** [comment]
and then https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory_client_access.go#L339 should likely be replaced with something like
```
args := ""
func parseFunc(flag *Flag, value string) error {
   args = args + " --" + flag.Name()
   if set, ok := flag.Annotations["classified"]; !ok || !set {
       args = args + "=" + value
   } else {
       args = args + "=" + CLASSIFIED
   }
}
pflag.ParseAll(parseFunc)
```

**Comment by xilabao** [comment]
@eparis  PTAL

**Comment by xilabao** [comment]
@eparis PTAL

**Comment by eparis** [comment]
@k8s-bot ok to test

**Comment by xilabao** [comment]
@eparis @lavalamp PTAL

**Comment by eparis** [comment]
/lgtm
@smarterclayton can you approve? changes in vendor require a top level person :-(

**Comment by xilabao** [comment]
> /lgtm
@smarterclayton can you approve? changes in vendor require a top level person :-(

Can someone approve this PR? @smarterclayton @lavalamp @dchen1107 @thockin

**Comment by rmmh** [comment]
@k8s-bot verify test this

**Comment by xilabao** [comment]
@k8s-bot verify test this

**Comment by xilabao** [comment]
@k8s-bot bazel test this

**Comment by xilabao** [comment]
fixed verify test @eparis PTAL

**Comment by xilabao** [comment]
fixed verify test  @smarterclayton @fabianofranz @eparis @lavalamp @rmmh PTAL

**Comment by liggitt** [comment]
I'll second @lavalamp, I don't see how this protects us, and it seems like a ton of effort to support a questionable feature.

**Comment by xilabao** [comment]
@liggitt the pr is to fix the problem in https://github.com/kubernetes/kubernetes/pull/35030

**Comment by xilabao** [comment]
@k8s-bot cvm gce e2e test this

**Comment by eparis** [comment]
lavalamp comment was concerning what was, I think, a completely different patch.  What is the ton of effort? This will hide things from the recorded command line. Doesn't seem like a waste to me...

**Pull request merged**